### PR TITLE
Change the interface of the PullRequest resource to be more granular.

### PR DESCRIPTION
**Pull request by dlorenc**
# Changes

Currently the PullRequest resource produces one large json object representing the
entire PR. This works, but makes it harder to write simple programs to manipulate this
object.

This commit changes this interface to produce a set of files, each representing a part
of the initial pull request. This makes it easier for programs to do simple tasks like add
labels or comments.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
 The pullRequest resource type now outputs and expects files in a slightly different format. The new format can be found [here](TODO).
```

### CODE REVIEW

File: cmd/pullrequest-init/disk_test.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+			{Text: "me"},
+		},
+	}
+
+	d, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(d)
+	fmt.Println(d)
```
Comment (bobcatfish) [inline cmd/pullrequest-init/disk_test.go]: included by accident?

File: cmd/pullrequest-init/disk_test.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	for _, c := range tektonPr.Comments {
+		actualComment, ok := comments[c.ID]
+		if !ok {
+			t.Errorf("Expected comment with ID: %d, not found: %v", c.ID, comments)
+		}
+		if diff := cmp.Diff(actualComment, *c); diff != "" {
+			t.Errorf("Get Comment: -want +got: %s", diff)
+		}
+	}
+}
```
Comment (bobcatfish) [inline cmd/pullrequest-init/disk_test.go]: the way this test and this code are broken up makes me think we could break up `ToDisk` even more into some separate methods with some separate tests - pretty much every section that's like "now we check x" "now we check y"

File: cmd/pullrequest-init/disk_test.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+
+func TestFromDisk(t *testing.T) {
+	d, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	// defer os.RemoveAll(d)
```
Comment (bobcatfish) [inline cmd/pullrequest-init/disk_test.go]: commented out by accident?

File: cmd/pullrequest-init/disk_test.go (hunk 2 above)
Comment (dlorenc) [reply cmd/pullrequest-init/disk_test.go]: Done. I don't actually think I like separating out the tests though.. The logic of each one is very simple, and if we tested at the focused level we'd still want to test at the overall level to make sure the overall function didn't forget to call one of the sub functions.

LMK what you think...

**Comment by dlorenc** [comment]
> What do you think if we also included something in `examples` that maybe just read from a PR in a public repo?

Yea, I'm going to add that one as a separate PR along with an integration test.

**Comment by dlorenc** [comment]
> I think there are a couple of doc changes that need to go along with this:
> 
> * https://github.com/tektoncd/pipeline/blob/master/docs/resources.md#pull-request-resource
> * https://github.com/tektoncd/pipeline/tree/master/cmd/pullrequest-init#pullrequest-init
>   (Feels like those 2 docs repeat each other a bit, so if you want the README to link to the resources doc that could be good - whatever you think is best!)

Done!

File: cmd/pullrequest-init/disk_test.go (hunk 2 above)
Comment (bobcatfish) [reply cmd/pullrequest-init/disk_test.go]: i agree with still having one test at the top level  - however if it was up to me id still make individual tests, then make the test of the top level function as simple as possible (maybe even just going to disk and back again, making sure some of the expected values were present). the approach im suggesting probably does have more potential to miss something, but i think in conjunction with an end to end test that actually uses this, we'd have pretty much everything covered.

i find individual tests much easier to read and maintain over time, and easier to make sure that everything is covered as desired (e.g. reading a small function + writing a test for it vs. reading a bunch of small functions while reading through one big test)

but i think we're getting very close to a situation where its just that i would have done it differently from you so im happy with it the way it is XD

if we need to start changing those individual functions them maybe we can revisit it

Comment (bobcatfish) [reply cmd/pullrequest-init/disk_test.go]: oh here's a thought tho: how do you plan to add in gitlab support? there could be some argument made for logic like so:

You could have `FromDisk` and `ToDisk` functions that do all of:
- comments
- labels
- statuses
- refs

But operate on different implementations of comments, labels, statuses and refs for each of GitHub and GitLab. In that case you might want some top level test that makes sure that FromDisk and ToDisk are operating correctly, but then separate unit tests for each of comments, labels, statuses, refs for each implementation?

I dunno just thoughts

File: cmd/pullrequest-init/README.md [hunk 4]
```diff
@@ -12,68 +12,7 @@ Currently supported providers:
...
-    https://developer.github.com/v3/repos/statuses/#get-the-combined-status-for-a-specific-ref
-*   `$PATH/github/comments/#.json`: Comments associated to the PR as specified
-    by https://developer.github.com/v3/issues/comments/#get-a-single-comment
-
-For now, these files are *read-only*.
-
-The binary will look for GitHub credentials in the `$GITHUBTOKEN` environment
-variable. This should generally be specified as a secret with the field name
-`githubToken` in the `PullRequestResource` definition.
+For information about the payloads written to disk, see the [resource documentation](../../docs.resources.md#pull-request-resource).
```
Comment (bobcatfish) [inline cmd/pullrequest-init/README.md]: wooooot

File: docs/resources.md [hunk 5]
```diff
@@ -136,69 +136,61 @@ Pull Request resource represents a pull request event from a source control
...
-            "Code": "success",
-            "Description": "Job succeeded.",
-            "ID": "pull-tekton-pipeline-go-coverage",
-            "URL": "https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/895/pull-tekton-pipeline-go-coverage/1141483806818045953/"
-        },
-  ],
-  "RawStatus": "/tmp/pr/github/status.json"
+  "Text": "comment body",
+  "Author": "author",
+  "ID": 202131633
```
Comment (bobcatfish) [inline docs/resources.md]: nice updates!

**Pull request merged**